qiita_pet/handlers/admin_processing_job.py

```diff
@@ -109,21 +110,43 @@ def post(self):
         # Get user-inputted qiita id and sample names
         qid = self.get_argument("qid")
         snames = self.get_argument("snames").split()
-        error, matching, missing, extra, blank = [None]*5
+        error, matching, missing, extra, blank, duplicates = [None]*6
 
         # Stripping leading qiita id from sample names
         # Example: 1.SKB1.640202 -> SKB1.640202
         try:
-            qsnames = list(Study(qid).sample_template)
+            st = Study(qid).sample_template
+            qsnames = list(st)
         except TypeError:
             error = f'Study {qid} seems to have no sample template'
         except QiitaDBUnknownIDError:
             error = f'Study {qid} does not exist'
 
         if error is None:
+            # if tube_id is present then this should take precedence in qsnames
+            tube_ids = dict()
+            if "tube_id" in st.categories:
+                for k, v in st.get_category("tube_id").items():
+                    # ignoring empty values
+                    if v in (None, 'None', ''):
+                        continue
+                    if k.startswith(qid):
+                        k = k.replace(f'{qid}.', "", 1)
+                    tube_ids[k] = v
+
             for i, qsname in enumerate(qsnames):
                 if qsname.startswith(qid):
-                    qsnames[i] = qsname.replace(f'{qid}.', "", 1)
+                    qsname = qsname.replace(f'{qid}.', "", 1)
```

Comment:
IMHO, I'm not a fan of big loops with lots of conditionals inside. I think the sanitizing of qsname and tube_ids could be handled better by list comprehensions before entering the loops. This would make the loops easier to read. However I understand this wasn't your code.

Reply:
I don't like multiple loops as things get slower and slower as things grow (remember we are dealing with GUI, users and lots of samples); thus, I prefer a single loop with a few conditionals inside.